sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala

```diff
@@ -128,6 +129,48 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
       TableIdentifier("tempTable"), ignoreIfNotExists = true, purge = false)
   }
 
+  test("SPARK-21079 - analyze table with location different than that of individual partitions") {
+    def queryTotalSize(tableName: String): BigInt =
+      spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes
+
+    withTempPaths(4) { (paths) => paths match {
+      case tablePath :: partitionPaths =>
+        sql(
+          s"""
+            |CREATE TABLE analyzeTable_part (key STRING, value STRING) PARTITIONED BY (ds STRING)
+            |LOCATION '${tablePath}'
+          """.stripMargin).collect()
+
+        val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
+        partitionDates.zip(partitionPaths).foreach(p => {
+          val ds = p._1
+          val path = p._2
+          sql(
+            s"""
+              |ALTER TABLE analyzeTable_part ADD PARTITION (ds='${ds}')
+              |LOCATION '${path.toString}'
+            """.stripMargin).collect()
+          sql(
+            s"""
+              |INSERT INTO TABLE analyzeTable_part PARTITION (ds='${ds}')
+              |SELECT * FROM src
+            """.
+              stripMargin).collect()
+          }
+        )
+
+        // Modify table location to not match location of individual partitions
+        sql("ALTER TABLE analyzeTable_part SET LOCATION 'file:/do/not/use'").collect()
+
+        sql("ANALYZE TABLE analyzeTable_part COMPUTE STATISTICS noscan")
+
+        assert(queryTotalSize("analyzeTable_part") === BigInt(17436))
+
+        sql("DROP TABLE analyzeTable_part").collect()
```

Comment:
use `withTable` to include the test, then we don't need to drop table explicitly.